protocol/buffer.go

```diff
@@ -534,31 +580,21 @@ func (ref *pageRef) ReadByte() (byte, error) {
 }
 
 func (ref *pageRef) Read(b []byte) (int, error) {
-	if ref.cursor >= int64(ref.length) {
-		return 0, io.EOF
-	}
 	n, err := ref.ReadAt(b, ref.cursor)
 	ref.cursor += int64(n)
 	return n, err
 }
 
 func (ref *pageRef) ReadAt(b []byte, off int64) (int, error) {
-	limit := ref.offset + int64(ref.length)
-	off += ref.offset
-
-	if off >= limit {
+	offset := int64(ref.offset)
+	length := int64(ref.length)
+	if off >= length {
 		return 0, io.EOF
 	}
-
-	if off+int64(len(b)) > limit {
-		b = b[:limit-off]
+	if off+int64(len(b)) > length {
+		b = b[:length-off]
```

Comment:
Is this needed because a pageRef could end partially through a page and we would potentially read past the end of the pageRef without this adjustment?

Reply:
Potentially. This is mostly to satisfy the `io.ReaderAt` interface I believe. I revisited the implementation to simplify it a bit when I was debugging.